### DISCUSSION TITLE
Triangle shape elements

### DISCUSSION THREAD

**Discussion by onajjar** [damogranlabs/classy_blocks#59] [General]
Hello, 

I am trying to generate an element with six points having an extruded triangle shape the same shape as the blockMesh website in the following link: 
https://www.openfoam.com/documentation/user-guide/4-mesh-generation-and-conversion/4.3-mesh-generation-with-the-blockmesh-utility
 
It is working perfectly as I want, as attached in the photo,  but I used blockMesh to generate it. However, I want to use classy_blocks since I am generating a much more complex mesh. I know it maybe an easy task but it seems that I couldn't solve it ! 

Thanks in advance 

![Capture d’écran du 2024-11-02 12-10-14](https://github.com/user-attachments/assets/210d9ddd-9325-4643-94e5-0e9c9224a4e5)

Omar

**Comment by FranzBangar** [comment]
Hello,

thanks for using this library!
I did use triangular blocks in the past myself but it never worked properly - in a coarse mesh it did but with the 'real' thing (finer case) everything exploded. I decided to drop non-quads altogether.

Although it is possible to support triangular blocks with a few minor modifications, I'd suggest you could maybe rethink your blocking to consist of quads only. Can you share your geometry (or at least a part of it?) Maybe I can help with that (?)

**Reply by onajjar** [reply]
Hello,



Thank you for your response.

I am attaching a .zip file that contains the following( test_crack.zip):

1. Four CSV files representing the x, y, and z coordinate matrices for points on two surfaces (top terrain surface and bottom terrain).
2. The main code file, "brazilian_DEAP_crack_test.py," which generates the crack volume using multiple boxes placed side by side.
3. "functions.py," which includes helper functions for finding points inside a circle.
4. "DEAP_crack_occ.py," which loads the data (I previously used it with Gmsh, but it didn't work as expected).

The goal is to create a cylindrical hole inside the crack, but I am uncertain where exactly the cylinder intersects the crack since the z elevation varies at each point. Currently, I am attempting to remove boxes that are entirely within a circle, then generate the remaining elements around the circle based on its position.

The idea that I had to use triangle shape elements is that when a circle crosses the box, it deletes 6 points out of 8, the shape will be an arc connected with a 2 straight sides (almost like the triangle element) (something like this in the photo) 
![image](https://github.com/user-attachments/assets/ca1efdbd-0bbc-46f2-9a20-b52ce4567ea3)

I would greatly appreciate your assistance with this!

Thank you.

**Comment by MHS0027** [comment]
I think a 8 node block
With a clipping plane that intersects two points on one edge will create
triangular shape in classy blocks

Regards Malcolm

**Reply by FranzBangar** [reply]
In The Times Of Yore, classy could produce wedges but now I'm not sure this is so. Chopping and Grading require edges of more than zero length...

**Comment by FranzBangar** [comment]
Your approach might be the most obvious and might seem the easiest; although I could add a check for zero-length edges to classy and it would produce a legit blockMesh with wedges but I see some issues:
1. You'd create triangular prisms; at two corners, blocks would be approximately square but at the third there would be very thin triangles. You'd have to decide at which side to merge two block points.
1. Some triangles would be okay (equilateral-ish) but some would become terribly thin.
1. Blocks that will sit next to triangles would have square cells but their immediate neighbours will be those thin wedges. I don't know what kind of simulation you'll do but that size difference is quite a party pooper in many cases!

I don't understand completely why you can't define where your cylinder intersects the crack. If you could define its radius and position, I think the simplest approach would be:
1. Use your `triangulate()` script to obtain two STL files for top and bottom of the crack
1. Calculate cylinder's position and domain boundaries
1. Create a simple ExtrudedShape from a Sketch with a hole in the middle - only a few blocks! (like this https://github.com/damogranlabs/classy_blocks/blob/development/showcase/karman.png but maybe even without the inner 'ring' as you don't need boundary inflation layers)
1. Project cylinder's inner surfaces to a `searchableCylinder`
1. Project all surfaces, edges and vertices of all blocks' bottom and top faces to bottom and top STLs.

What do you think? It seems too easy so I suspect I overlooked an important detail.

**Reply by onajjar** [reply]
Indeed, there would be some issues regarding the sizes of elements around the cylinder as you said. 

I have tried a similar to your approach in gmsh, and I succeeded in importing the mesh in the CFX program, as the photo shows,
![image](https://github.com/user-attachments/assets/2ff34d4e-c50d-4f59-8d09-82d29e0fc97d)
,  but the simulation gives an overflow fatal error, which I believe is due to the mesh quality.  

It seems, indeed, a much easier approach, but since I am only using classy for a couple of days, it will not be direct for me to know. 

May I ask if I give you the two STL files for the top and bottom surfaces to help me with it, please? 

Thanks again for your suggestions.

**Reply by FranzBangar** [reply]
Yes, I'd gladly try that. Wonder where it'll go wrong.

**Reply by onajjar** [reply]
Thanks a lot ! you will end my suffer ! 

Please, find the two stl files in the following link : 
https://filesender.renater.fr/?s=download&token=f5e7d536-1a3c-48b3-8247-510a6938e98b

**Comment by FranzBangar** [comment]
Of course it's not as straightforward as I hallucinated. It's a short piece of code (currently without the cylinder part but it's not important at the moment):

```python
#!/usr/bin/env python
import classy_blocks as cb

mesh = cb.Mesh()

cell_size = 0.001

xmin = -0.055
xmax = 0.055
ymax = 0.05
zmin = -0.01
zmax = 0.01

face = cb.Face([[xmin, 0, zmin], [xmax, 0, zmin], [xmax, ymax, zmin], [xmin, ymax, zmin]])
extrude = cb.Extrude(face, 0.01)

extrude.chop(0, start_size=cell_size)
extrude.chop(1, start_size=cell_size)
extrude.chop(2, count=5)

extrude.project_side('bottom', 'min', edges=True, points=True)
extrude.project_side('top', 'max',  edges=True, points=True)

mesh.add(extrude)

mesh.add_geometry({
    "max": [
        "type triSurfaceMesh",
        "file surface_max.stl",
    ],
    "min": [
        "type triSurfaceMesh",
        "file surface_min.stl",
    ]
})


mesh.write('system/blockMeshDict')
```

But it doesn't work very well.
![image](https://github.com/user-attachments/assets/25e9f754-3568-4360-a1b9-029169db544a)
You can see that wherever there are sharp-ish creases the projections create garbage, intersecting bottom and top surfaces etc. However, this is just a single block. I'll try creating a grid of blocks and will let you know if that helps.

**Reply by onajjar** [reply]
It is already not bad! Indeed, it is a much easier approach and saves a lot of lines! I hope it will work with cylinders

**Reply by onajjar** [reply]
I have changed the distance between the two surfaces to a large distance; the results are always the same ! Maybe it takes one surface.

**Comment by FranzBangar** [comment]
This is made with a great bunch of blocks but nothing is projected (doesn't work apparently):
![image](https://github.com/user-attachments/assets/b3828f8b-6e15-4286-97b3-105893ca3a0e)

A coarser mesh is okay but `checkMesh` already reports high skewness with this, finer mesh. And this is uniform thickness ('snapped' to one surface, the other has just a fixed offset).
How precisely do you need to mesh these surfaces?

**Reply by onajjar** [reply]
My simulation is airflow from one side to another along the y direction. So, I want to have at least 25 elements in the z-direction ( with a boundary layer) and at least 100 elements in the other two directions ( x and y). Also, a boundary layer towards the cylinder space. 

The distance between the two surfaces is indeed very small and it is not constant (which is the idea that I want to illustrate). I know it will give bad elements quality, but for the moment, I have no choice since the distance is very small compared to the structure size.

**Reply by FranzBangar** [reply]
I'm guessing maybe snappyHexMesh would be better suited for this? Did you give it a try already?

**Reply by FranzBangar** [reply]
I guess not. Don't try that at home unless you have jigabytes of RAM and bazillion processors.
![image](https://github.com/user-attachments/assets/4af89e4a-1320-481f-8450-8f35f0304296)

**Reply by onajjar** [reply]
I looked at it, and I did not manage to do it.

**Comment by FranzBangar** [comment]
So I assume your approach is the right one, you just have to correct the blocking of the cylinder.
Instead of deleting points that fall within the circle, remove the whole block (so that you end up with a square hole), then make blocks between a straight side of the hole and the cylinder:
![image](https://github.com/user-attachments/assets/7f7bc722-4574-4a93-9ed6-fb3f68f8b1c8)

As for the projections and mesh quality: I'd make curves from your CSV points, then put block edges on them. I'll try some of that tomorrow.

**Reply by onajjar** [reply]
Careful consideration needs to be taken regarding the z elevation for the cylinder segments. We can obtain them by interpolating the z values of the deleted blocks. Indeed, it could work.

**Reply by onajjar** [reply]
Finally it worked perfectly using your idea ! 
Thank you a lot !

**Reply by FranzBangar** [reply]
Great job, I'm glad it worked. As of today, I was still fiddling with various attempts. Obviously you have given this much more thought, hence, success.

**Reply by onajjar** [reply]
I'm testing the code on different surfaces (different .csv files) without holes, but sometimes, I encounter this issue:

****--> FOAM FATAL IO ERROR: (openfoam-2106)
Block hex (17 16 20 21 19 18 22 23) (8 6 50) simpleGrading (1(1) 1(1) 
2
(
(0.5 0.5 79.4968)
(0.5 0.5 0.0125791)
)
) has inward-pointing faces
    4(17 21 20 16)****
    
It happens for some of csv files. Do you have any ideas or suggestions on what might be causing this problem?

Thanks again!

**Reply by FranzBangar** [reply]
Is it possible that your surfaces intersect?
If you're sure that's not the case, it might be that the blocking (which points you choose and which you ignore) produce self-intersecting blocks. You can check that by opening `debug.vtk` in paraview (see [this tutorial](https://damogranlabs.com/2023/09/classy_blocks-tutorial-part-5-some-useful-hints/)).
Do you use straight edges or do you already have curves in your blocks?

**Reply by onajjar** [reply]
The two surfaces never intersect because  I have put a minimum distance between them as a criteria. 

I have fixed the problem but in a wierd way. It worked when I removed the mean surface elevation ! 

I couldn't understand why it worked and why it is important for blockMesh to know what is the mean z elevation for the points ( does it make difference when I mesh the geometry on z_mean=0 or z_mean= 1 .etc?) ! 
I moved a mean value from the top surface and the same mean from the bottom surface so the distance between the points did not change and it worked ! 

Now, I have another issue which maybe handled using the tutorial you have sent me. I need to extrude a patch and not only one face. Since I am adding many blocks, I have defined a patch face for the boundary I want and now, I want to extrude this patch.  Is it possible?

**Reply by FranzBangar** [reply]
This is a weird solution and might be a bug in classy_blocks. I'll see if I can think of anything.

If you need something like "thicken surface" or "offset" in CAD programs, you might use `Shell` ([this example](https://github.com/damogranlabs/classy_blocks/blob/development/examples/shape/shell.py)) - you give it a list of Faces and it will create 'thickened' blocks. You might need to correct the offset points though... Let me know if that's what you need.

**Reply by onajjar** [reply]
I am not sure if it is a bug in classy_blocks or inside blockMesh itself. I have seen this link, https://bugs.openfoam.org/view.php?id=3349, facing almost the same problem in blockMesh. This check about inward point face can be discarded by adding checkBlockFaceOrientation false; to blockMeshDict but I did not do it since it is also important to check the orientation.

**Reply by FranzBangar** [reply]
Well those checks by blockMesh might not be in order (some are done before creating actual cells and don't take into account edges and projections). In any case, you should always make sure checkMesh says the mesh is okay. If blockMesh finds errors, the mesh might still be okay but a failed checkMesh won't run a simulation.
